File: api/app/domain/utils/safe_redirect.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from urllib.parse import urlparse


def resolve_safe_redirect_path(redirect: str | None, *, default: str = "/") -> str:
    """Return a same-origin relative path for post-login redirects."""
    candidate = (redirect or "").strip()
    if not candidate:
        return default
    if not candidate.startswith("/"):
        return default
    if candidate.startswith("//"):
        return default
    parsed = urlparse(candidate)
    if parsed.scheme or parsed.netloc:
        return default
    return candidate

File: ui/src/lib/safe-redirect.ts
const DEFAULT_REDIRECT = "/";

export function resolveSafeRedirectPath(redirect: string | null | undefined): string {
  const candidate = (redirect ?? "").trim();
  if (!candidate) {
    return DEFAULT_REDIRECT;
  }
  if (!candidate.startsWith("/") || candidate.startsWith("//")) {
    return DEFAULT_REDIRECT;
  }
  try {
    const parsed = new URL(candidate, "http://localhost");
    if (parsed.origin !== "http://localhost") {
      return DEFAULT_REDIRECT;
    }
    return `${parsed.pathname}${parsed.search}${parsed.hash}`;
  } catch {
    return DEFAULT_REDIRECT;
  }
}
